### PR TITLE
Combine adversary play with FCP population

### PR DESCRIPTION
**Pull request by avv-va**
This PR combines adversary play + FCP population:
1. Train a population of agent with originaler learner type
2. Select the best agent out of the population
3. Create an adversary(selfisher) from random policy to attack the best agent
4. Train an agent from random policy with teammates of [itself, adversary], [itself, high], [itself, medium] (assuming N=3, it can use curriculum or not depending on the config)
5. Repeat step 3 and 4 for the number of attacks

It also cleans some of the existing functions, notably:
1. args variable is **only** being set in `train_helper.py` in `set_input` function and before calling set_input() in if \_\_name\_\_ == \_\_main\_\_. 
2. if a variable is specific to a function, it should only be set in the specific function for that method for example attack_rounds is specific to adversaries so it's only being set in N_X_SP_w_adversaries and is not set in args. But, total_train_timesteps is something that is meaningful for all the methods, so it's being set in args in set_input.

### CODE REVIEW

File: scripts/train_agents.py [hunk 1]
```diff
@@ -418,21 +417,52 @@ def N_X_SP(args,
...
         )
 
 
+def N_X_SP_w_adversaries(args) -> None:
+
+    args.unseen_teammates_len = 1
+    args.primary_train_types = [TeamType.SELF_PLAY_HIGH, TeamType.SELF_PLAY_HIGH, TeamType.SELF_PLAY_HIGH, TeamType.SELF_PLAY_HIGH,
+                                TeamType.SELF_PLAY_MEDIUM, TeamType.SELF_PLAY_MEDIUM, TeamType.SELF_PLAY_MEDIUM, TeamType.SELF_PLAY_MEDIUM,
+                                TeamType.SELF_PLAY_ADVERSARY]
+
```
Comment (Jmiceli17) [inline scripts/train_agents.py]: Why are you repeating `TeamTypes` here? What is that doing?

File: scripts/utils/train_helper.py [hunk 2]
```diff
@@ -248,15 +367,15 @@ def get_adversary(args, total_training_timesteps, train_types, eval_types, curri
                          h_dim=args.ADV_h_dim, 
                          train_types=train_types,
                          has_curriculum= not curriculum.is_random)
-    agent = load_agent(Path(agent_path), args)
```
Comment (Jmiceli17) [inline scripts/utils/train_helper.py]: why was this removed? I'm also confused how the call to `generate_TC_for_Adversary()` is working

**Review by Jmiceli17 — COMMENTED** [review]
Looks good. Only major question is in `train_helper.py`

File: scripts/train_agents.py (hunk 1 above)
Comment (avv-va) [reply scripts/train_agents.py]: It adds more than high agent in the teammates_collection

`tc[self_play_high] = [[random, high], [random, high], ...]`

File: scripts/utils/train_helper.py (hunk 2 above)
Comment (avv-va) [reply scripts/utils/train_helper.py]: this was a mistake but I don't use this get_adversary function. I use `get_adversary_agent()`

File: scripts/train_agents.py (hunk 1 above)
Comment (Ttopiac) [reply scripts/train_agents.py]: If we don’t use a predefined curriculum, then in a 2-agent game, the teammate combinations will allocate 4/9 of the time to [self, high], 4/9 of the time to [self, medium], and 1/9 of the time to [self, adv].

Is this understanding correct? Please feel free to correct me if needed. Thanks!

File: scripts/utils/train_helper.py [hunk 3]
```diff
@@ -1,131 +1,258 @@
...
+    else:
+        dont_attack_N_X_SP(args=args,
+                           population=population,
+                           curriculum=curriculum,
+                           unseen_teammates_len=unseen_teammates_len,
+                           n_x_sp_eval_types=n_x_sp_eval_types
+                           )
+
+
+def attack_N_X_SP(args, population, curriculum, unseen_teammates_len, adversary_play_config, attack_rounds, n_x_sp_eval_types):
```
Comment (narendasan) [inline scripts/utils/train_helper.py]: Is this to train a new agent against adversaries in the set of teammates? Would it be possible to have one function that does both given a population? Then you could make more diverse population sets instead of adv vs standard?

Comment (avv-va) [reply scripts/utils/train_helper.py]: I am not sure if I understand  your question, could you clarify it? The difference between adv and standard is that, adversaries are trained to attack the agent in an iterative manner:
1. train adv0 to target the primary0 agent
2. train primary1 with adv + pop
3. train new adv1 that targets primary1
4. train primary2 with adv1 + adv0 + pop
5. repeat

whereas the standard (i.e the population) could be any agent with any policy that is created once and is not created iteratively

Comment (narendasan) [reply scripts/utils/train_helper.py]: So is the goal of this function to train the adversary?

Comment (narendasan) [reply scripts/utils/train_helper.py]: or is the this like a joint adversary / ego agent training loop?

File: oai_agents/gym_environments/base_overcooked_env.py [hunk 4]
```diff
@@ -34,7 +34,7 @@
 class OvercookedGymEnv(Env):
     metadata = {'render.modes': ['human']}
 
-    def __init__(self, grid_shape=None, ret_completed_subtasks=False, stack_frames=False, is_eval_env=False,
+    def __init__(self, learner_type, grid_shape=None, ret_completed_subtasks=False, stack_frames=False, is_eval_env=False,
```
Comment (Ttopiac) [inline oai_agents/gym_environments/base_overcooked_env.py:37]: if leaner_type does not have a default, we cannot run run_overcooked_game properly
suggested changes:
`def __init__(self, learner_type=LearnerType.ORIGINALER, grid_shape=None, ret_completed_subtasks=False, stack_frames=False, is_eval_env=False, shape_rewards=False, enc_fn=None, full_init=True, args=None, num_enc_channels=27, deterministic=False, **kwargs):`

File: scripts/utils/train_helper.py (hunk 3 above)
Comment (avv-va) [reply scripts/utils/train_helper.py]: yes the latter: train adv + train ego with adv -> repeat for attack_rounds

File: oai_agents/agents/rl.py [hunk 5]
```diff
@@ -260,9 +265,25 @@ def should_evaluate(self, steps):
...
+        self.curriculum.print_curriculum()
+        print("How Long: ", self.args.how_long)
+        print(f"Epoch timesteps: {self.epoch_timesteps}")
+        print(f"Total training timesteps: {total_train_timesteps}")
+        print(f"Number of environments: {self.n_envs}")
+        print(f"Hidden dimension: {self.hidden_dim}")
+        print(f"Seed: {self.seed}")
+        print(f"Checkpoint rate: {self.checkpoint_rate}")
+        print(f"Learner type: {self.learner_type}")
+        print("Dynamic Reward? ", self.args.dynamic_reward)
```
Comment (Ttopiac) [inline oai_agents/agents/rl.py]: is the question mark"?" a typo in print("Dynamic Reward? ", self.args.dynamic_reward)?

suggested edits:
`print("Dynamic Reward: ", self.args.dynamic_reward)`

**Review by Ttopiac — COMMENTED** [review]
good work! I only left only two comments, required small changes.

File: oai_agents/agents/rl.py (hunk 5 above)
Comment (avv-va) [reply oai_agents/agents/rl.py]: fixed!

File: oai_agents/gym_environments/base_overcooked_env.py (hunk 4 above)
Comment (avv-va) [reply oai_agents/gym_environments/base_overcooked_env.py:37]: I fixed this by passing a custom learner_type in `overcooked_gui.py`. your suggested change would work but it's a risky fix because if we ever forget to pass the learner type when training the agents it would not run into an error and instead would run with the default value and we would not know that we have made a mistake.

File: scripts/utils/train_helper.py (hunk 3 above)
Comment (narendasan) [reply scripts/utils/train_helper.py]: ah, I think the names tripped me up. Maybe something like `joint_adv_N_X_SP` and `N_X_SP` might make sense since attack/dont_attack makes it seem like some sort of post ego-agent training function

File: scripts/train_agents.py [hunk 6]
```diff
@@ -453,40 +458,67 @@ def N_X_SP(args,
...
+
+def N_1_SP(args) -> None:
     '''
     The randomly initialized agent will train with itself and one other unseen teammate (e.g. [SP, SP, SP, SP_H] in a 4-chef layout)
     
     :param pop_force_training: Boolean that, if true, indicates population should be generated, otherwise load it from file
     :param primary_force_training: Boolean that, if true, indicates the SP agent teammates_collection should be trained  instead of loaded from file
-    :param parallel: Boolean indicating if parallel envs should be used for training or not
+    :param  : Boolean indicating if   envs should be used for training or not
```
Comment (Jmiceli17) [inline scripts/train_agents.py]: We can delete these `:param` comments since the function signature no longer includes them

File: scripts/utils/train_helper.py (hunk 3 above)
Comment (Jmiceli17) [reply scripts/utils/train_helper.py]: I agree with Naren that the naming is a little confusing
